### ISSUE TITLE
Move from web.py to something else

### ISSUE THREAD

**Issue by charmander** [Weasyl/weasyl#5]
Anything else. Flask?

**Comment by Syfaro** [comment]
Flask has always been lovely to work with, it's by far the web framework I'm most comfortable with. I've written stuff like my multiuploader with it.

Apparently some really large sites are using it too (Pinterest, Uber, few others).

**Comment by weykent** [comment]
Flask is utter garbage, especially for large web applications.

Pyramid is a good choice, and I hear there's already been some work done with it.

**Comment by waweber** [comment]
I've used Pyramid in a few applications. It's convenient, especially if you're writing an API. In our case, the most beneficial features would be its URL traversal scheme and how it handles permissions/access control.

**Comment by strain-113** [comment]
URL traversal is actually the only controversial point about it right now: It made the initial attempt at Weasyl3 somewhat awkward. Between URL traversal and the way we were doing ORM it created a lot of strange inefficiencies (e.g. fetching a user object in order to fetch a single submission because of how we structured our resources and URL hierarchy).

I don't think that these issues are inevitable (and some of the difficulties may have merely been our unfamiliarity). So traversal may be worth revisiting (we discussed it a bit in Gitter). But I think the ordering right now is going to go:

Templates in Jinja2 -> Replace web.py core with pyramids with URL dispatch -> Consider traversal

**Comment by strain-113** [comment]
To be clear: I don't think Pyramid's URL Traversal is bad (although I did find it counter-intuitive last time). I think we should get the project running on Pyramid first and then consider restructuring things around traversal so adoption isn't blocked on rewriting our controller architecture.

Thoughts?

**Comment by weykent** [comment]
I think the way weasyl3 did traversal was a failed experiment. I think I discussed it before on how it _should_ be done, but only briefly, and I never wrote up the code. I don't think it's worth getting into too much detail right now, but I think the _right_ way to handle traversal is to not do any database queries during traversal, but just roll all of the path segments into an object that represents all of the access checks that need to be made, and do them all at once _after_ traversal.

Anyway. I still think the right way forward in doing an incremental port is:

- Make a wrapper around pyramid that looks like web.py, so the existing web.py code can use pyramid without realizing it. This means that new code can use pyramid features if desired. Pyramid can handle being passed rendered html, so there's no need to update the templates first. This means that the transition isn't blocked on getting _every_ template updated.
- Port web.py's routing so that it's usable from pyramid. There aren't actually very many routes, so maybe it would be easier to just convert them all to pyramid style non-traversal match routing.
- Port the templates to jinja2, while also porting view functions to pyramid as needed.

**Comment by charmander** [comment]
One point of moving away from web.py is to use Python 3, which is blocked on rewriting all the templates anyway. Since it’s relatively easy to rewrite templates in Jinja one at a time without changing frameworks, I’d suggest starting with that. There’s less setup work involved.

**Comment by weykent** [comment]
I don't know why python 3 would even be a milestone or target. If anything, pypy is a more logical target, since pypy is the future of python. I haven't tried recently to see what parts of the stack don't have cffi ports and don't run well under cpyext.

**Comment by strain-113** [comment]
I'd like to move away from the argument about whether it makes sense to do the templates first or swap out the core engine first, as they can be done fairly well in parallel.

I'll take at porting to Pyramid (with URL dispatch) this coming weekend.

**Comment by waweber** [comment]
I wrote a (very sparse) example in a branch in weasyl3 on traversal. Each traversal segment doesn't fetch any info from the database, but it might (optionally) check if the given resource exists in the case of a GET query.

I don't think I added it, but the __acl__ attribute for a submission resource, for example, should be a descriptor which will fetch the relevant information and return the appropriate ACL entries, for example a submission with rating 40 might grant the "view" permission to those with the "can_view_explicit" principal. This won't (i think) be called until a context has been found and a view is being looked up

**Comment by charmander** [comment]
@weykent: Python 3 is a significantly better language than Python 2. PyPy would be faster than CPython on a project where Python execution speed is not our bottleneck.

**Comment by strain-113** [comment]
There are architectural reasons to prefer PyPy that go beyond speed, but I think that discussion is a little out of scope here. Regardless of Python 3 support, there are reasons to prefer moving away from web.py.

**Comment by strain-113** [comment]
This is coming along gradually: https://github.com/strain-113/weasyl/tree/pyramid See https://github.com/strain-113/weasyl/blob/pyramid/pyramid.md in particular.

Status highlights:
 * I have the middleware layers seemingly working, 404 and general errors seem to be doing the right thing (can't test sentry).
 * `web` has been removed from define.py for everything except rendering the templates. We're using thread locals for most other things:
    -- `web.ctx` general maps to `request`. `web.ctx.env` is `request.environ`.
    -- Most other fields we put on `web.ctx` have been replaced with either pyramid lazily loaded properties or attributes.
 * The index view loads. So that's something.

Overall it's fine, although debugging this is going to be challenging.

At this stage I think the next step is porting over all the controllers. Looking at urls.py, there's about 200 of these. I plan to rename 'urls.py' to 'routes.py' and set it up as a centralized place to set up all routes for the application. I'm not sure whether to centralize setting up the views here with config.add_view() or using the `@view_config()` decorate approach with `config.scan()`. The latter seems more pyramid idiomatic and will be pithier for endpoints that handle GET and POST separately, but I think it may make it harder for a reader to find which code handles a particular URL (it's really just as difficult as a search, but it's going to mean looking at a file and then searching as opposed to just looking at a file). Thoughts?

More importantly:

The controllers existing controllers are all web.py style views, that is classes with GET and/or PUT methods. There are about 200 of these. To port these I can see one of two techniques:
 * The minimum approach that changes as little as possible would be to register these controller classes with `config.add_view()` or decorators directly. GET/POST methods can be handled with the `request_method` param to `add_view()`. Our controller_base sets up some logic with login_required that change the definite of GET and PUT at request time to handle errors. I've tested this and even with `@view_config()` decorators, it seems to do the right thing. We'll still have to handle using `request.params` instead of `web.input()`.
 * The other approach is to try to rewrite these all in a pyramid idiomatic way. I'm not sure what this would look like. Possibly `login_required`, `guest_required`, `moderator_only`, `admin_only`, and `disallow_api` would all become function decorators and we use functions for each of the old GET or POST views. 

Not sure how I feel about this. One mindset is 'Do the former because the most important thing is getting off web.py as soon as possible. We can do cleanup incrementally.' The other mindset is 'Getting onto pyramid is stupid if we're just going to write code that looks like web.py anyway. Clean it up or it never gets cleaned up.'

Thoughts? If we favor porting things without rewriting the code more than necessary for now, what would we do for a theoretical new endpoint?

**Comment by weykent** [comment]
:+1: on rewrite it all. It doesn't have to be manual, either—there's a few projects around that make it pretty straightforward to rewrite python code. All of the view modules are simple enough that it would be easy to do a one-shot conversion, I think.

As for how to structure the code, I agree with function (not class) views at the top level of the modules. The various `_required` etc. things can be parameters to `view_config`.

**Comment by strain-113** [comment]
I'm skeptical of anything like automates rewriting. So you have any examples?

If we do do automation what should be done about form = web.input()? We could write a new method that builds a new map out of request.params in the same fashion.

**Comment by weykent** [comment]
`2to3` from the python standard library is a generic framework for refactoring and rewriting. I believe there's tools based on it which exist separately.

**Comment by hyena** [comment]
https://github.com/hyena/weasyl/tree/pyramid now has the index, /search, /popular, and /streaming seemingly working.

I'll be porting others gradually, but it's a good time to critique the overall approach.

**Comment by hyena** [comment]
Opened #96. A pull request is more appropriate to track technical minutia arguments.
